### PR TITLE
[WIP] record the maximum memory used by LSF jobs

### PR DESCRIPTION
**Pull request by drkennetz**
Taking over @mr-c PR #2857 . This will currently log the command and max mem used by the command for an LSF job utilizing `bjobs -l`:

```
[2020-11-12T16:08:40-0600] [Thread-3  ] [I] [toil.batchSystems.lsf] [job ID 111913722, Command  _toil_worker CWLJob file:/research/rgs01/home/clusterHome/ dkennetz/toil_test/cwl/TEST kind-CWLJob/instance-shr9enqg --context gANjdG9pbC5iYXRjaFN5c3RlbXMuYWJzdHJhY3RCYXRjaFN5 c3RlbQpXb3JrZXJDbGVhbnVwQ29udGV4dApxACmBcQF9cQIoWBEAAAB3b3 JrZXJDbGVhbnVwSW5mb3EDY3RvaWwuYmF0Y2hTeXN0ZW1zLmFic3RyYWN0 QmF0Y2hTeXN0ZW0KV29ya2VyQ2xlYW51cEluZm8KcQRYPwAAAC9yZXNlYX JjaC9yZ3MwMS9ob21lL2NsdXN0ZXJIb21lL2RrZW5uZXR6L3RvaWxfdGVz dC9jd2wvd29ya2RpcnEFWCQAAAAxNWI5Mzg0Yi02NGJiLTQ0ODYtODFmZi 02YWU5MTI4MWFlYjlxBlgGAAAAYWx3YXlzcQeHcQiBcQlYBQAAAGFyZW5h cQpOdWIu] the maximum memory used was: 1 Gbytes
[2020-11-12T16:08:41-0600] [Thread-3  ] [I] [toil.batchSystems.lsf] [job ID 111913722, Command  _toil_worker CWLJob file:/research/rgs01/home/clusterHome/ dkennetz/toil_test/cwl/TEST kind-CWLJob/instance-shr9enqg --context gANjdG9pbC5iYXRjaFN5c3RlbXMuYWJzdHJhY3RCYXRjaFN5 c3RlbQpXb3JrZXJDbGVhbnVwQ29udGV4dApxACmBcQF9cQIoWBEAAAB3b3 JrZXJDbGVhbnVwSW5mb3EDY3RvaWwuYmF0Y2hTeXN0ZW1zLmFic3RyYWN0 QmF0Y2hTeXN0ZW0KV29ya2VyQ2xlYW51cEluZm8KcQRYPwAAAC9yZXNlYX JjaC9yZ3MwMS9ob21lL2NsdXN0ZXJIb21lL2RrZW5uZXR6L3RvaWxfdGVz dC9jd2wvd29ya2RpcnEFWCQAAAAxNWI5Mzg0Yi02NGJiLTQ0ODYtODFmZi 02YWU5MTI4MWFlYjlxBlgGAAAAYWx3YXlzcQeHcQiBcQlYBQAAAGFyZW5h cQpOdWIu] the maximum memory used was: 5.1 Gbytes
[2020-11-12T16:08:42-0600] [Thread-3  ] [I] [toil.batchSystems.lsf] [job ID 111913722, Command  _toil_worker CWLJob file:/research/rgs01/home/clusterHome/ dkennetz/toil_test/cwl/TEST kind-CWLJob/instance-shr9enqg --context gANjdG9pbC5iYXRjaFN5c3RlbXMuYWJzdHJhY3RCYXRjaFN5 c3RlbQpXb3JrZXJDbGVhbnVwQ29udGV4dApxACmBcQF9cQIoWBEAAAB3b3 JrZXJDbGVhbnVwSW5mb3EDY3RvaWwuYmF0Y2hTeXN0ZW1zLmFic3RyYWN0 QmF0Y2hTeXN0ZW0KV29ya2VyQ2xlYW51cEluZm8KcQRYPwAAAC9yZXNlYX JjaC9yZ3MwMS9ob21lL2NsdXN0ZXJIb21lL2RrZW5uZXR6L3RvaWxfdGVz dC9jd2wvd29ya2RpcnEFWCQAAAAxNWI5Mzg0Yi02NGJiLTQ0ODYtODFmZi 02YWU5MTI4MWFlYjlxBlgGAAAAYWx3YXlzcQeHcQiBcQlYBQAAAGFyZW5h cQpOdWIu] the maximum memory used was: 5.1 Gbytes
```

### CODE REVIEW

**Comment by drkennetz** [comment]
The random spaces in paths and text were introduced by copying.

File: src/toil/batchSystems/lsf.py [hunk 1]
```diff
@@ -297,6 +298,35 @@ def parseBjobs(self,bjobs_output_str):
...
+            memargs = ["bjobs", "-l", str(jobID)]
+            try:
+                bjobs = subprocess.check_output(memargs, universal_newlines=True)
+                memregex = r"MAX MEM: (.*?);"
+                meminfo = re.search(memregex, bjobs)
+                s = " ".join(bjobs.split())
+                command = re.search(r"Command <(.*?)>", s)
+                if meminfo:
+                    if not command:
+                        logger.info("Missing Command: %s", bjobs)
```
Comment (DailyDreaming) [inline src/toil/batchSystems/lsf.py:317]: Consider a more verbose: `Cannot Parse Max Memory Due to Missing Command String`.

**Review by DailyDreaming — APPROVED** [review]
@drkennetz Is this still WIP?  I'm approving and would be fine with merging as is.

One comment is below and also the `parseMaxMem` function doesn't seem to need to return anything, so I would consider removing the `return` statements.

**Comment by drkennetz** [comment]
@DailyDreaming I gave it a return because @mr-c said it might be useful to actually return from getJobExitCode in the future for provenance. It serves no purpose at the moment though, except as one less thing to think about if we decide to implement this change in the future.

I am happy to remove the return, and make the logging more verbose as suggested. It is no longer a WIP. I'll wait on your response before committing changes (potentially removing the return and updating logging).

File: src/toil/batchSystems/lsf.py [hunk 2]
```diff
@@ -297,6 +298,35 @@ def parseBjobs(self,bjobs_output_str):
...
+                        logger.info("[job ID %s, Command %s] the maximum memory used was: %s",
+                                    str(jobID), command.group(1), meminfo.group(1))
+                else:
+                    logger.debug("[job ID %s] Unable to collect maximum memory usage: %s",
+                                 str(jobID), bjobs)
+                return meminfo
+            except subprocess.CalledProcessError as err:
+                logger.debug("[job ID %s] Unable to collect maximum memory usage: %s",
+                             str(jobID), str(err))
+                return 0
```
Comment (mr-c) [inline src/toil/batchSystems/lsf.py:328]:
```suggestion
            return None
```
Makes this a good default return

**Review by mr-c — CHANGES_REQUESTED** [review]

**Comment by drkennetz** [comment]
Sorry for the delay, I have updated based on the recommendations here.

**Review by mr-c — APPROVED** [review]

**Pull request merged**